src/torch/SIRF_torch.py

```diff
@@ -0,0 +1,269 @@
+        #  [batch, *src_shape]
+
+        # check the src shape
+        torch_src_shape = torch_src.shape
+        if len(torch_src_shape) == len(self.sirf_src_shape):
+            raise ValueError(f"Invalid shape of src. Expected batch dim.")
+        elif len(torch_src_shape) == len(self.sirf_src_shape) + 1:
+            check_shapes(torch_src_shape[1:], self.sirf_src_shape)
+            if self.sirf_src_shape == torch_src_shape[1:]:
+                torch_src = torch_src.unsqueeze(1) # add channel dimension
+                self.channel = True
+        elif len(torch_src_shape) == self.sirf_src_shape + 2:
+            check_shapes(torch_src_shape[2:], self.sirf_src_shape)
+
+        n_batch = torch_src.shape[0]
+        n_channel = torch_src.shape[1]
+
+        # This looks horrible, but PyTorch will be able to trace.
+        batch_dest = []
+        for batch in range(n_batch):
+            channel_dest = []
+            for channel in range(n_channel):
+                channel_dest.append(self._operator.apply(torch_src[batch, channel], 
+                    self.sirf_src_template, self.operator))
+            batch_dest.append(torch.stack(channel_dest, dim=0))
+        if self.channel:
+            return torch.stack(batch_dest, dim=0).squeeze(1)
+        # [batch, channel, *sirf_dest.shape]
+        return torch.stack(batch_dest, dim=0)
+
+
+class SIRFObjectiveFunction(torch.nn.Module):
+    def __init__(self,
+            sirf_obj_func: object,
+            sirf_image_template: object
+            ):
+        super(SIRFObjectiveFunction, self).__init__()
+        self.sirf_obj_func = sirf_obj_func
+        self.sirf_image_template = sirf_image_template
+        self.sirf_image_shape = sirf_image_template.shape
+
+    def forward(self, torch_image):
+        # PyTorch src is size [batch, channel, *sirf_image_shape] or
+        #  [batch, *sirf_image_shape]
+
+        torch_image_shape = torch_image.shape
+        if len(torch_image_shape) == len(self.sirf_image_shape):
+            raise ValueError(f"Invalid shape of src. Expected batch dim.")
+        elif len(torch_image_shape) == len(self.sirf_image_shape) + 1:
+            check_shapes(torch_image_shape[1:], self.sirf_image_shape)
+            if self.sirf_image_shape == torch_image_shape[1:]:
+                torch_image = torch_image.unsqueeze(1) # add channel dimension
+                self.channel = True
+        elif len(torch_image_shape) == len(self.sirf_image_shape) + 2:
+            check_shapes(torch_image_shape[2:], self.sirf_image_shape)
+
+        n_batch = torch_image.shape[0]
+        n_channel = torch_image.shape[1]
+
+        # This looks horrible, but PyTorch should be able to trace.
+        batch_values = []
+        for batch in range(n_batch):
+            channel_values = []
+            for channel in range(n_channel):
+                channel_values.append(_ObjectiveFunction.apply(torch_image[batch, channel], self.sirf_image_template, self.sirf_obj_func))
```

Comment:
@KrisThielemans  we need to update SIRF's objective functions "on the fly", we could have an input of multiple objective functions?